streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java

```diff
@@ -195,6 +224,33 @@ public static void cleanStateAfterTest(final EmbeddedKafkaCluster cluster, final
         IntegrationTestUtils.produceKeyValuesSynchronouslyWithTimestamp(topic, records, producerConfig, timestamp, false);
     }
 
+    /**
+     * @param topic               Kafka topic to write the data records to
+     * @param records             Data records to write to Kafka
+     * @param producerConfig      Kafka producer configuration
+     * @param headers             {@link Headers} of the data records
+     * @param timestamp           Timestamp of the record
+     * @param <K>                 Key type of the data records
+     * @param <V>                 Value type of the data records
+     */
+    public static <K, V> void produceKeyValuesSynchronouslyWithTimestamp(final String topic,
+                                                                         final Collection<KeyValue<K, V>> records,
+                                                                         final Properties producerConfig,
+                                                                         final Headers headers,
+                                                                         final Long timestamp)
```

Comment:
Was there a specific reason for adding this method?

Reply:
For parity. `produceKeyValuesSynchronouslyWithTimestamp` has 3 required and 3 optional parameters[^1]. By adding this method (overload 2), we can provide a counterpart of overload 4. (see below) Additionally, I found that this method is handier than using overload 4 for header parameters.

|          | headers | timestamp | enableTransactions |
|----------|---------|-----------|--------------------|
|     1    |    X    |     O     |          X         |
| 2 (new!) |    O    |     O     |          X         |
|     3    |    X    |     O     |          O         |
|     4    |    O    |     O     |          O         |

[^1]: topic, records, producerConfig / headers, timestamp, enableTransactions.

Reply:
That makes sense, but is this method currently unused?

If it's not used, then I think it's better not to add it. (IMHO, lack of dead code outweighs the value of symmetry)

Reply:
@vvcephei Thank you for your comment. Yes, you are right - this method is not used in the project. But `IntegrationTestUtils` is also intended to be provided to the end users [as a testing utility for Kafka Streams applications, with `EmbeddedKafkaCluster`.](https://github.com/dongjinleekr/kafka-streams-example/blob/master/src/test/java/com/dongjinlee/examples/kafkastreams/IntegrationTest.java)

For this reason, I thought adding this method is reasonable, with documentation for the end users.

Reply:
Thanks for the response, @dongjinleekr .

I'm not too sure if `IntegrationTestUtils` is _intended_ to be a public API, or if it's just been used that way, since there was no good way to write integration tests without `EmbeddedKafkaCluster`. I think that example is more of a symptom of how strong the need is, with no good solution available.

I've heard people mention in the past that there's some desire to offer a real public integration test api, but I couldn't find any evidence of it on Jira. I've proposed https://issues.apache.org/jira/browse/KAFKA-7852 , so maybe we can do something about it.

Regardless, generally speaking, a project's own tests shouldn't be considered a public API. So, while I see your point about the de facto state of things, it still doesn't seem quite right to start adding methods specifically for public use.

Plus there a catch-22: if it's _not_ a public API, then we can make any change we want in a PR, but if it _is_ a public API, then we must go through a KIP to propose this change ;)

Reply:
As far as I know, `IntegrationTestUtils` and `EmbeddedKafkaCluster` are intended to be used as a testing utility. Here is an excerpt from [Kafka Streams in Action](https://www.manning.com/books/kafka-streams-in-action) by @bbejeck. As you can see, both classes are used for integration tests for Kafka Streams application.

<img width="556" alt="streams-test" src="https://user-images.githubusercontent.com/2375128/51590931-fde34f00-1f2e-11e9-946a-a06ae23d0b25.png">

However, I totally agree that there is no good way for testing Kafka Streams applications, in an end user's perspective. I experienced several questions about how to write testing code from the Kafka communities in Seoul and Tokyo. (In fact, the example above was originally written to show them.) @vvcephei, You are right - this problem reflects the strong need for testing utility, not only violates the principle that "a project's own tests shouldn't be considered a public API" or not have to be exposed for public use.

It seems like there was a plan to develop the testing utility module, but what I could find was [a draft KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-139%3A+Kafka+TestKit+library) only. I think this topic should be discussed after 2.2.0 release.

So, in short, how about this?

1. Remove this new overload (i.e., keep the status quo.)
2. Merge [KAFKA-7852](https://issues.apache.org/jira/browse/KAFKA-7852) into [KAFKA-5041](https://issues.apache.org/jira/browse/KAFKA-5041) by @ijuma (As @mjsax pointed out, it is a duplication.)
3. Continue to discuss this topic with the KIP later.